gitpod-io/gitpod
# [bridge] Update prebuild status when controlling instance timeouts (5/5)

## Description

This PR fixes a bug with prebuilds being stuck in `queued` indefinitely in rare cases. (5/5 in the chain)

This turned out to be bigger then anticipated, because the `MetaInstanceController` we had in place beforehand was ill-conceived, and did only work for self-hosted installations. With this change it's part of `Bridge`, so going to be executed per workspace cluster, so it's going to run in production as well (as well as Self-Hosted, of course).

For documenting this change we now:
 - have a [diagram](https://www.notion.so/gitpod/WebApp-Architecture-784cf14bc76440e3beb8de56f61e7340#6c9bd05eef5a44408171a95b14ae73a9)
 - [updated description](https://www.notion.so/gitpod/WebApp-Architecture-784cf14bc76440e3beb8de56f61e7340#ed5196fd4fdb47319c42ec4a781d232c) for `BridgeController` and `Bridge` reflecting this change


## Related Issue(s)

Fixes: https://github.com/gitpod-io/gitpod/issues/9395, #8274
Depends on: https://github.com/gitpod-io/gitpod/pull/10727

## How to test
 - start a [workspace on this branch](https://gitpod.io/#https://github.com/gitpod-io/gitpod/pull/10426) and `./dev/preview/install-k3s-kubeconfig.sh`
 - `kubectl edit cm ws-manager-bridge-config` and make the timeout for the `building` phase `60`s
 - [start a prebuild](https://gpl-9395-sccfe07d632.preview.gitpod-dev.com/workspaces) that is depending on an image build that takes longer than 1min
   - for instance: clone [this branch](https://github.com/geropl/gitpod-test-repo/blob/gpl/docker-build-10mins/.gitpod/Dockerfile)
   - create a project for it
   - update that Dockerfile
 - check how workspace and prebuild get updated and marked as stopped/aborted in the DB
   - watch the ws-manager-bridge logs for: "Marking workspace instance as stopped"
   - that log line contains the workspaceId of the prebuild workspace
 - check the DB tables `d_b_workspace_instance` and `d_b_prebuilt_workspace` to both show that those are stopped/aborted :heavy_check_mark: 

## Release Notes

```release-note
fix prebuilds stuck in `queued` indefinitely
```

## Documentation
